resources/aws.go

```diff
@@ -0,0 +1,334 @@
+type AwsRes struct {
+	BaseRes `yaml:",inline"`
+	State   string `yaml:"state"` // state: running, stopped, terminated
+	Region  string `yaml:"region"`
+	Type    string `yaml:"type"`
+	ImageID string `yaml:"imageid"`
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *AwsRes) Default() Res {
+	return &AwsRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *AwsRes) Validate() error {
+	if obj.State != "running" && obj.State != "stopped" && obj.State != "terminated" {
+		return fmt.Errorf("State must be 'running', 'stopped' or 'terminated'")
+	}
+	return obj.BaseRes.Validate()
+}
+
+// Init initializes the resource.
+func (obj *AwsRes) Init() error {
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *AwsRes) Watch() error {
+	send := false
+	var exit *error
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region: aws.String(obj.Region),
+	}))
+	svc := ec2.New(sess)
+	diInput := ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String(obj.GetName())},
+			},
+			{
+				Name:   aws.String("instance-state-name"),
+				Values: []*string{aws.String("running"), aws.String("pending"), aws.String("stopped"), aws.String("stopping")},
+			},
+		},
+	}
+	if err := obj.Running(); err != nil {
+		return err
+	}
+	awschan := make(chan string)
+	go func() {
+		for {
+			diOutput, err := svc.DescribeInstances(&diInput)
+			if err != nil {
+				errwrap.Wrapf(err, "error describing instances")
+			}
+			var waitInput ec2.DescribeInstancesInput
+			if len(diOutput.Reservations) == 1 {
+				waitInput = ec2.DescribeInstancesInput{
+					InstanceIds: []*string{diOutput.Reservations[0].Instances[0].InstanceId},
+				}
+			} else {
+				waitInput = diInput
+			}
+
+			if obj.State == "running" && len(diOutput.Reservations) == 1 {
+				svc.WaitUntilInstanceStopped(&waitInput)
+				if *diOutput.Reservations[0].Instances[0].State.Name != "running" {
+					awschan <- "stopped"
+					svc.WaitUntilInstanceRunning(&waitInput)
+				}
+			}
+			if obj.State == "stopped" && len(diOutput.Reservations) == 1 {
+				svc.WaitUntilInstanceRunning(&waitInput)
+				if *diOutput.Reservations[0].Instances[0].State.Name != "stopped" {
+					awschan <- "running"
+					svc.WaitUntilInstanceStopped(&waitInput)
+				}
+			}
+			if obj.State == "terminated" {
+				svc.WaitUntilInstanceExists(&waitInput)
+				if len(diOutput.Reservations) != 0 {
+					awschan <- "exists"
+					svc.WaitUntilInstanceTerminated(&waitInput)
+				}
+			}
+
+		}
+	}()
+	for {
+		select {
+		case event := <-obj.Events():
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit
+			}
+		case state := <-awschan:
+			log.Printf(state)
+			obj.StateOK(false)
+			send = true
+			if send {
```

Comment:
this send should be after the select because it could trigger from obj.Events() as well... see other resources